File: payload/extensions-samples/spring-cloud-stream.extract.py
"""Sample seam extractor (Python): Spring Cloud Stream functional bindings -> msg_edges."""
import re

BEAN_RE = re.compile(r"@Bean\s+(?:public\s+)?(Consumer|Supplier|Function)\s*<[^>]*>\s+(\w+)\s*\("
                     r"|fun\s+(\w+)\s*\(\s*\)\s*:\s*(Consumer|Supplier|Function)\b")


def _kafka(text, ctx):
    cfg = ctx["config"]
    out = []
    for m in BEAN_RE.finditer(text):
        kind = (m.group(1) or m.group(4) or "").lower()
        name = m.group(2) or m.group(3)
        if not name:
            continue
        line = text.count("\n", 0, m.start()) + 1
        in_key = f"spring.cloud.stream.bindings.{name}-in-0.destination"
        out_key = f"spring.cloud.stream.bindings.{name}-out-0.destination"
        if kind in ("consumer", "function") and cfg.get(in_key):
            out.append({"topic": cfg[in_key], "direction": "consume", "line": line, "via": in_key})
        if kind in ("supplier", "function") and cfg.get(out_key):
            out.append({"topic": cfg[out_key], "direction": "produce", "line": line, "via": out_key})
    return out


extractors = {"kafka": _kafka}

File: payload/extensions-samples/spring-cloud-stream.extract.mjs
/** Sample seam extractor: Spring Cloud Stream functional bindings.
 * @Bean Consumer/Supplier/Function beans + spring.cloud.stream.bindings.<name>-in/out-0.destination
 * become msg_edges, flowing through message_flow/docgen like native Kafka edges.
 * Drop into .ariadne/extensions/, the same contract covers gateways, gRPC, GraphQL, etc. */
const BEAN_RE = /@Bean\s+(?:public\s+)?(Consumer|Supplier|Function)\s*<[^>]*>\s+(\w+)\s*\(|fun\s+(\w+)\s*\(\s*\)\s*:\s*(Consumer|Supplier|Function)\b/g;

export const extractors = {
  kafka(text, { configMap, relpath }) {
    const cfg = (k) => (configMap?.get ? configMap.get(k) : configMap?.[k]);
    const out = [];
    const lineAt = (i) => text.slice(0, i).split("\n").length;
    for (const m of text.matchAll(BEAN_RE)) {
      const kind = (m[1] ?? m[4] ?? "").toLowerCase();
      const name = m[2] ?? m[3];
      if (!name) continue;
      const line = lineAt(m.index);
      const inKey = `spring.cloud.stream.bindings.${name}-in-0.destination`;
      const outKey = `spring.cloud.stream.bindings.${name}-out-0.destination`;
      if ((kind === "consumer" || kind === "function") && cfg(inKey)) {
        out.push({ topic: cfg(inKey), direction: "consume", line, via: inKey });
      }
      if ((kind === "supplier" || kind === "function") && cfg(outKey)) {
        out.push({ topic: cfg(outKey), direction: "produce", line, via: outKey });
      }
    }
    return out;
  },
};
